[RFC] implement convergence control in MIR using SelectionDAG

ssahasra opened the pull request:
LLVM function calls carry convergence control tokens as operand bundles, where the tokens themselves are produced by convergence control intrinsics. This patch implements convergence control tokens in MIR as follows:

1. Introduce target-independent ISD opcodes and MIR opcodes for convergence control intrinsics.
2. Model token values as untyped virtual registers in MIR.

The actual lowering of controlled convergent operations (including function calls) and convergence control intrinsics is target-specific.

On AMDGPU, the convergence control operand bundle at a non-intrinsic call is translated to an explicit argument to the SI_CALL_ISEL instruction. Post-selection adjustment converts this explicit argument to an implicit argument on the SI_CALL instruction. For intrinsics, AMDGPU introduces an AMDGPUISD opcode CONVERGENCECTRL_GLUE and a corresponding machine opcode with the same spelling. This is used as a glued argument to SDNodes that is later translated to an implicit argument in the MIR.

All convergent intrinsics on AMDGPU need to set hasPostISelHook in their description.

ssahasra commented:
#67006 is a slightly outdated attempt at doing the same thing with GlobalISel. That other PR will be redesigned after completing this current PR.

ssahasra commented:
Also, forgot to mention that this is a very early RFC, hoping to attract attention to the general scheme of things. The code contains occasional debug code, and the implementation is incomplete (for example, SelectINTRINSIC_W_CHAIN is not implemented, and the LOOP intrinsic is not implemented).

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp:
```diff
@@ -2627,7 +2637,7 @@ void AMDGPUDAGToDAGISel::SelectINTRINSIC_WO_CHAIN(SDNode *N) {
   case Intrinsic::amdgcn_interp_p1_f16:
     SelectInterpP1F16(N);
-    return;
+    break;
```
This isn't convergent so I don't see why this would need to change

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp:
```diff
@@ -2610,7 +2610,17 @@ void AMDGPUDAGToDAGISel::SelectINTRINSIC_W_CHAIN(SDNode *N) {
 void AMDGPUDAGToDAGISel::SelectINTRINSIC_WO_CHAIN(SDNode *N) {
   unsigned IntrID = cast<ConstantSDNode>(N->getOperand(0))->getZExtValue();
-  unsigned Opcode;
+  unsigned Opcode = 0;
```
technically 0 is a valid opcode, should use INSTRUCTION_LIST_END

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -901,6 +901,20 @@ bool AMDGPUTargetLowering::aggressivelyPreferBuildVectorSources(EVT VecVT) const
+void AMDGPUTargetLowering::CollectTargetIntrinsicOperands(
+    const CallInst &I, SmallVectorImpl<SDValue> &Ops, SelectionDAG &DAG,
+    function_ref<SDValue(const Value *)> getValue) const {
+  if (auto Bundle = I.getOperandBundle(LLVMContext::OB_convergencectrl)) {
```
There's nothing AMDGPU specific here. This is general infrastructure so I would expect the target callback to be unnecessary

ssahasra replied:
The new SDNode being returned has the AMD-specific GLUE opcode. It's not clear to me that we can decide for all targets whether the LLVM convergence operand bundle should be translated to a glued operand. So this allows each target to return whatever node fits their scheme.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -5969,6 +5970,24 @@ bool SelectionDAGBuilder::visitEntryValueDbgValue(const DbgValueInst &DI) {
+    setValue(&I, DAG.getNode(ISD::CONVERGENCECTRL_ENTRY, sdl, MVT::Untyped));
+    break;
+  case Intrinsic::experimental_convergence_loop: {
+    assert(false && "coming soon");
```
llvm_unreachable

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -8351,7 +8382,8 @@ void SelectionDAGBuilder::LowerCallTo(const CallBase &CB, SDValue Callee,
           CB.countOperandBundlesOfType(LLVMContext::OB_preallocated) != 0)
-      .setCFIType(CFIType);
+      .setCFIType(CFIType)
+      .setConvergenceControlToken(ConvControlToken);
```
Skip creating the undef if it's not a convergent call?

ssahasra replied:
The TD pattern for SI_CALL_ISEL always expects a non-null token, and it is applied independent of whether the call is convergent.

ssahasra reviewed (commented):
@arsenm do you see any potential pitfalls in how the glue node is being passed around? The glue is needed only on intrinsics, and there are a couple of FIXME comments about the possibility of some intrinsic having multiple glue nodes. Also, at one point, SelectNodeTo() loses the glue node, so this change first stashes the node, then calls SelectNodeTo(), and then adds the glue node back with MorphNodeTo().

arsenm commented:
> @arsenm do you see any potential pitfalls in how the glue node is being passed around? The glue is needed only on intrinsics, and there are a couple of FIXME comments about the possibility of some intrinsic having multiple glue nodes. Also, at one point, SelectNodeTo() loses the glue node, so this change first stashes the node, then calls SelectNodeTo(), and then adds the glue node back with MorphNodeTo().

This seems like just a bug in SelectNodeTo. Glue shouldn't just get lost, that kind of defeats the purpose

nhaehnle commented:
Our of curiosity, what's happening with this?

ssahasra commented:
> Our of curiosity, what's happening with this?

It keeps getting pushed back behind other tasks!

ssahasra commented:
> > @arsenm do you see any potential pitfalls in how the glue node is being passed around? The glue is needed only on intrinsics, and there are a couple of FIXME comments about the possibility of some intrinsic having multiple glue nodes. Also, at one point, SelectNodeTo() loses the glue node, so this change first stashes the node, then calls SelectNodeTo(), and then adds the glue node back with MorphNodeTo().
> 
> This seems like just a bug in SelectNodeTo. Glue shouldn't just get lost, that kind of defeats the purpose

I gave the wrong impression earlier. The problem is ``SelectCode``, which is generated from TD patterns. ``AMDGPUTargetLowering::CollectTargetIntrinsicOperands()`` appends an extra CONVERGENCECTRL_GLUE operand to the node, which is ignored by the generated selection code. Doing it right will require adding the token operand to every convergent opcode in our TD files. The easier hack is to just restore the token using ``MorphNodeTo()`` after calling ``SelectCode()``.

Or there is a very real possibility that I misunderstood how glue nodes are used?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -245,6 +245,13 @@ static cl::opt<bool, true> LateCFGStructurize(
+static cl::opt<bool, true> DisableStructurizer(
+    "amdgpu-disable-structurizer",
+    cl::desc("Disable structurizer for experiments; produces unusable code"),
+    cl::location(AMDGPUTargetMachine::DisableStructurizer), cl::Hidden);
```
Maybe make ReallyHidden

arsenm commented on llvm/test/CodeGen/AMDGPU/verify-convergencectrl/basic.mir:
```diff
@@ -0,0 +1,37 @@
+# RUN: not --crash llc -mtriple=amdgcn-- -run-pass=machineverifier -o /dev/null %s 2>&1 | FileCheck %s
```
This test should go in test/MachineVerifier, and can use -run-pass=none

arsenm commented on llvm/test/CodeGen/AMDGPU/verify-convergencectrl/cycles.mir:
```diff
@@ -0,0 +1,52 @@
+# RUN: not --crash llc -mtriple=amdgcn-- -run-pass=machineverifier -o /dev/null %s 2>&1 | FileCheck %s
```
Ditto

arsenm commented on llvm/test/CodeGen/AMDGPU/verify-convergencectrl/mixed2.mir:
```diff
@@ -0,0 +1,15 @@
+# RUN: not --crash llc -mtriple=amdgcn-- -run-pass=machineverifier -o /dev/null %s 2>&1 | FileCheck %s
```
Ditto

arsenm commented on llvm/test/CodeGen/AMDGPU/verify-convergencectrl/not-ssa.mir:
```diff
@@ -0,0 +1,11 @@
+# RUN: not --crash llc -mtriple=amdgcn-- -run-pass=machineverifier -o /dev/null %s 2>&1 | FileCheck %s
```
Ditto

arsenm commented on llvm/test/CodeGen/AMDGPU/verify-convergencectrl/region-nesting.mir:
```diff
@@ -0,0 +1,24 @@
+# RUN: not --crash llc -mtriple=amdgcn-- -run-pass=machineverifier -o /dev/null %s 2>&1 | FileCheck %s
```
Ditto

ssahasra commented:
Squashed and rebased. Fixed the lit tests to use "-run-pass=none". The current state seems good enough to drop the "RFC" tag.

ssahasra commented:
bump!

ssahasra commented:
Bump! The only TODO is to mark every convergent function with HasPostISelHook = 1 in the TD files for AMDGPU. I am working on cleaner ways of doing this, but it need not block this review. Having the actual implementation of MIR tokens in upstream will be very useful.

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -14775,6 +14798,21 @@ void SITargetLowering::AdjustInstrPostInstrSelection(MachineInstr &MI,
+
+  if (SDNode *GluedNode = Node->getGluedNode()) {
+    // FIXME: Possibly iterate over multiple glue nodes?
+    GluedNode->dump();
```
Leftover debug printing

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -14775,6 +14798,21 @@ void SITargetLowering::AdjustInstrPostInstrSelection(MachineInstr &MI,
+    if (GluedNode->getOpcode() == ~(unsigned)AMDGPU::CONVERGENCECTRL_GLUE) {
+      Register VReg = getVR(GluedNode->getOperand(0));
+      MachineOperand MO = MachineOperand::CreateReg(VReg, /* isDef = */ false,
+                                                    /* isImp = */ true);
```
No space before = to be recognized by clang-format

arsenm commented on llvm/lib/Target/AMDGPU/SIInstructions.td:
```diff
@@ -629,10 +629,15 @@ def SI_CALL_ISEL : SPseudoInstSI <
+  (SI_CALL_ISEL $src0, (i64 0), untyped:$token)
 >;
 
+// def : GCNPat<
+//   (AMDGPUcall i64:$src0, (i64 0)),
+//   (SI_CALL_ISEL $src0, (i64 0), undef_tied_input)
+// >;
```
Commented out code

arsenm commented on llvm/lib/CodeGen/SelectionDAG/FunctionLoweringInfo.cpp:
```diff
@@ -394,6 +394,16 @@ Register FunctionLoweringInfo::CreateRegs(const Value *V) {
                                       !TLI->requiresUniformRegister(*MF, V));
 }
 
+Register FunctionLoweringInfo::InitializeRegForValue(const Value *V) {
```
Why did you move this out of line?

ssahasra replied:
Because `isa<ConvergenceControlInst>(V)` needs to include IntrinsicInst.h ...

arsenm commented on llvm/lib/CodeGen/SelectionDAG/InstrEmitter.cpp:
```diff
@@ -1193,7 +1193,9 @@ EmitMachineNode(SDNode *Node, bool IsClone, bool IsCloned,
-    TLI->AdjustInstrPostInstrSelection(*MIB, Node);
+    TLI->AdjustInstrPostInstrSelection(
+        *MIB, Node,
+        [this, &VRBaseMap](SDValue Op) { return getVR(Op, VRBaseMap); });
```
Can you avoid this new callback parameter by adding a map for the tokens to FunctionLoweringInfo?

ssahasra replied:
Does that mean that I should pass FLI as a parameter here instead of the lambda? Because TLI does not have a pointer to FLI ...

ssahasra replied:
Actually, the `InstEmitter::getVR()` method invoked inside that callback does extra work like creating a new reg for every use of an implicit-def. So the current design looks like the right way to me, unless I am missing something.

arsenm replied:
Yes, I mean pass around FunctionLoweringInfo. 

The implicit_def special casing looks weird to me. Seems like an unnecessary pre-optimization for something that's easy to rematerialize

arsenm replied:
The implicit_def special case shouldn't be relevant here. It should be illegal to have IMPLICIT_DEF define a token

arsenm replied:
Do we really need the target implementation to deal with this? Can we make the assumption that a token operand is the final MI operand and just have InstrEmitter directly handle it?

ssahasra replied:
Actually, the reason I went down this path is that `CONVERGENCECTRL_GLUE` is an AMDGPU opcode. I had started by thinking that glueing the token operand to an `SDNode` is a target-specific hack. But at this point, I am fairly confident that this scheme is suitable for all targets. That will simplify other places too, such as the callback passed to `AMDGPUTargetLowering::CollectTargetIntrinsicOperands()`.

As far as I can see, the token still has to be glued, and it eventually becomes an implicit use in the MIR. Otherwise a lot of convergent operations will need to be updated in the TD files.

ssahasra commented:
Pushed a commit that elevates CONVERGENCECTRL_GLUE to be a target-independent opcode. This is much simpler because handling tokens no longer needs target callbacks and also does not need post-isel adjustment hook on convergent operations.

arsenm commented on llvm/include/llvm/ADT/GenericConvergenceVerifier.h:
```diff
@@ -32,11 +32,12 @@ template <typename ContextT> class GenericConvergenceVerifier {
   void initialize(raw_ostream *OS,
                   function_ref<void(const Twine &Message)> FailureCB,
-                  const FunctionT &F) {
+                  const FunctionT &F, bool TokensAllowed_) {
```
Don't really see why this needs to be configurable?

arsenm replied:
I think the IsSSA would be a more sensible name

arsenm commented on llvm/lib/CodeGen/SelectionDAG/FunctionLoweringInfo.cpp:
```diff
@@ -394,6 +394,16 @@ Register FunctionLoweringInfo::CreateRegs(const Value *V) {
+  if (V->getType()->isTokenTy() && !isa<ConvergenceControlInst>(V))
+    return 0;
+  Register &R = ValueMap[V];
+  assert(R == 0 && "Already initialized this value register!");
```
R == Register()

arsenm commented on llvm/lib/CodeGen/SelectionDAG/InstrEmitter.cpp:
```diff
@@ -1191,6 +1191,21 @@ EmitMachineNode(SDNode *Node, bool IsClone, bool IsCloned,
     }
   }
 
+  if (Opc == TargetOpcode::CONVERGENCECTRL_GLUE) {
+    MIB->getOperand(0).setIsKill(false);
+  }
```
Why do you need this? General you use clearKillFlags but I wouldn't expect that to be necessary in InstrEmitter

ssahasra replied:
I actually don't remember why I put it. Removed!

arsenm reviewed (commented):
This is much simpler, a couple of nits

ssahasra commented on llvm/lib/CodeGen/SelectionDAG/InstrEmitter.cpp:
Line 342 is where the token operand gets killed on the GLUE node. The GLUE node is the only visible use of the token, because the actual use in READFIRSTLANE is this glue node.

ssahasra replied:
At line 342, I could check for "getGluedUser()" in general, or check for a use that has opcode CONVERGENCECTRL_GLUE. That seems clearer than resetting the killed flag later. Would it be wrong to do the general check for getGluedUser() instead of a specific opcode?

arsenm replied:
In theory we should be able to just drop all the attempts at setting kill flags in InstrEmitter. They're fully recomputed later and deprecated

arsenm commented on llvm/lib/CodeGen/SelectionDAG/InstrEmitter.cpp:
```diff
@@ -285,6 +285,27 @@ Register InstrEmitter::getVR(SDValue Op,
 }
 
+static bool isConvergenceCtrlIntrinsic(SDValue Op) {
+  if (Op->isMachineOpcode()) {
```
The not-machine opcode case should never reach InstrEmitter

ssahasra replied:
The emitter encounters ISD::CopyFromReg here. I am not so sure about making a special case for it. At least this way, the body of the function matches its name!

ssahasra replied:
@arsenm the review already has your approval, but I don't want to submit without closing all comments, especially this discussion about handling non-machine opcodes in instr-emitter.

arsenm replied:
CopyFromReg is fine, but you shouldn't need to handle the pre-selection convergence opcodes

ssahasra committed:
Implement convergence control in MIR using SelectionDAG

LLVM function calls carry convergence control tokens as operand bundles, where
the tokens themselves are produced by convergence control intrinsics. This patch
implements convergence control tokens in MIR as follows:

1. Introduce target-independent ISD opcodes and MIR opcodes for convergence
   control intrinsics.
2. Model token values as untyped virtual registers in MIR.

The change also introduces an additional ISD opcode CONVERGENCECTRL_GLUE and a
corresponding machine opcode with the same spelling. This glues the convergence
control token to SDNodes that represent calls to intrinsics. The glued token is
later translated to an implicit argument in the MIR.

The lowering of calls to user-defined functions is target-specific. On AMDGPU,
the convergence control operand bundle at a non-intrinsic call is translated to
an explicit argument to the SI_CALL_ISEL instruction. Post-selection adjustment
converts this explicit argument to an implicit argument on the SI_CALL
instruction.

ssahasra commented:
- squashed and rebased
- reworded the commit description to reflect current state

hctim commented:
Hi Sameer,

Looks like this broke the sanitizer buildbots: https://lab.llvm.org/buildbot/#/builders/5/builds/41499

```
FAIL: LLVM :: MachineVerifier/verify-implicit-def.mir (2 of 81037)
******************** TEST 'LLVM :: MachineVerifier/verify-implicit-def.mir' FAILED ********************
Exit Code: 1
Command Output (stderr):
--
RUN: at line 2: not --crash /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 -run-pass=none -o /dev/null /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/MachineVerifier/verify-implicit-def.mir 2>&1 | /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/FileCheck /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/MachineVerifier/verify-implicit-def.mir
+ not --crash /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 -run-pass=none -o /dev/null /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/MachineVerifier/verify-implicit-def.mir
+ /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/FileCheck /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/MachineVerifier/verify-implicit-def.mir
--
********************
Testing: 
FAIL: LLVM :: MachineVerifier/test_g_dyn_stackalloc.mir (3 of 81037)
******************** TEST 'LLVM :: MachineVerifier/test_g_dyn_stackalloc.mir' FAILED ********************
Exit Code: 1
Command Output (stderr):
--
RUN: at line 1: not --crash /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/llc -mtriple=aarch64 -o /dev/null -run-pass=none -verify-machineinstrs /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/MachineVerifier/test_g_dyn_stackalloc.mir 2>&1 | /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/FileCheck /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/MachineVerifier/test_g_dyn_stackalloc.mir
+ not --crash /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/llc -mtriple=aarch64 -o /dev/null -run-pass=none -verify-machineinstrs /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/MachineVerifier/test_g_dyn_stackalloc.mir
+ /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/FileCheck /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/MachineVerifier/test_g_dyn_stackalloc.mir

```

Looks like there's a memory corruption bug that can be revealed by removing the `-o /dev/null` from the tests.

You can reproduce the bots in full by using the instructions at https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild, but that's not fully necessary in this case. A quick repro is:

```
$ cmake \
-DCMAKE_C_COMPILER=clang \
-DCMAKE_CXX_COMPILER=clang++ \
-DLLVM_USE_LINKER=lld \
-GNinja \
-DCMAKE_BUILD_TYPE=Release \
-DCMAKE_C_FLAGS="-fsanitize=address" \
-DCMAKE_CXX_FLAGS="-fsanitize=address" \
-DLLVM_USE_SANITIZER=Address \
/path/to/llvm

$ LIT_OPTS='--filter MachineVerifier' ninja check-llvm

< ... snip ... >
********************                                                                                                      
********************                                                                                                      
Failed Tests (2):                                                                                                         
  LLVM :: MachineVerifier/test_g_dyn_stackalloc.mir                                                                       
  LLVM :: MachineVerifier/verify-implicit-def.mir                                                                         
```

ssahasra commented:
[AMD Official Use Only - General]

Thanks for the headsup, Mitch!

It's late pm in India, and I don't expect to have this fixed during the day tomorrow. If this is a blocker, can someone please revert the patch?

Thanks!
Sameer.

________________________________
From: Mitch Phillips ***@***.***>
Sent: 04 March 2024 21:38
To: llvm/llvm-project ***@***.***>
Cc: Sahasrabuddhe, Sameer ***@***.***>; State change ***@***.***>
Subject: Re: [llvm/llvm-project] [RFC] implement convergence control in MIR using SelectionDAG (PR #71785)

Caution: This message originated from an External Source. Use proper caution when opening attachments, clicking links, or responding.


Hi Sameer,

Looks like this broke the sanitizer buildbots: https://lab.llvm.org/buildbot/#/builders/5/builds/41499

FAIL: LLVM :: MachineVerifier/verify-implicit-def.mir (2 of 81037)
******************** TEST 'LLVM :: MachineVerifier/verify-implicit-def.mir' FAILED ********************
Exit Code: 1
Command Output (stderr):
--
RUN: at line 2: not --crash /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 -run-pass=none -o /dev/null /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/MachineVerifier/verify-implicit-def.mir 2>&1 | /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/FileCheck /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/MachineVerifier/verify-implicit-def.mir
+ not --crash /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 -run-pass=none -o /dev/null /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/MachineVerifier/verify-implicit-def.mir
+ /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/FileCheck /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/MachineVerifier/verify-implicit-def.mir
--
********************
Testing:
FAIL: LLVM :: MachineVerifier/test_g_dyn_stackalloc.mir (3 of 81037)
******************** TEST 'LLVM :: MachineVerifier/test_g_dyn_stackalloc.mir' FAILED ********************
Exit Code: 1
Command Output (stderr):
--
RUN: at line 1: not --crash /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/llc -mtriple=aarch64 -o /dev/null -run-pass=none -verify-machineinstrs /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/MachineVerifier/test_g_dyn_stackalloc.mir 2>&1 | /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/FileCheck /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/MachineVerifier/test_g_dyn_stackalloc.mir
+ not --crash /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/llc -mtriple=aarch64 -o /dev/null -run-pass=none -verify-machineinstrs /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/MachineVerifier/test_g_dyn_stackalloc.mir
+ /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/FileCheck /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/MachineVerifier/test_g_dyn_stackalloc.mir



Looks like there's a memory corruption bug that can be revealed by removing the -o /dev/null from the tests.

You can reproduce the bots in full by using the instructions at https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild, but that's not fully necessary in this case. A quick repro is:

$ cmake \
-DCMAKE_C_COMPILER=clang \
-DCMAKE_CXX_COMPILER=clang++ \
-DLLVM_USE_LINKER=lld \
-GNinja \
-DCMAKE_BUILD_TYPE=Release \
-DCMAKE_C_FLAGS="-fsanitize=address" \
-DCMAKE_CXX_FLAGS="-fsanitize=address" \
-DLLVM_USE_SANITIZER=Address \
/path/to/llvm

$ LIT_OPTS='--filter MachineVerifier' ninja check-llvm

< ... snip ... >
********************
********************
Failed Tests (2):
  LLVM :: MachineVerifier/test_g_dyn_stackalloc.mir
  LLVM :: MachineVerifier/verify-implicit-def.mir


—
Reply to this email directly, view it on GitHub<https://github.com/llvm/llvm-project/pull/71785#issuecomment-1976932608>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AJ53IZM7ZLEVV7K5BE4KXMLYWSMAPAVCNFSM6AAAAAA7EDJEUOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZWHEZTENRQHA>.
You are receiving this because you modified the open/close state.
